### PR TITLE
Fix transient nodes bug (really: fix bad handling of missing preimages in MIPS Preimage)

### PR DESCRIPTION
**Pull request by norswap**
WIP for fixing https://github.com/ethereum-optimism/cannon/issues/89

Currently this fails `make test_minigeth` from the Cannon repo, relevant part of the output:

```
# block 13284053 (deletion)
minigeth/go-ethereum 13284053
committed transactions 0x4b32fa3468c72852e5595025a2d8012a2f5112e04f9b9bd85f46e727e06ca4e1 <nil>
processing state: 13284053 -> 13284054
read 23 transactions
made block, parent: 0x5690741178538a1726722a433c4233cc179b5591363bfc895433ed22f56f9982
.......................consensus finalize
can't find preimage 0x9e978fd69bda076d4bece8f7cf7b6dd684d6cfb095a76f9779240ccd98254a68
new Root 0xcab252de675efe88ffe3b55de3e0ce14252ec7627bc746ce2211ce2ea7a81d20
receipt count 23 hash 0xda02504ad0cb3a614d2db9319fa966febceb2f0ab7d43530ac2656c4ecffd968
process done with hash 0x5cdfbdf585f5a3b3138d48d93abc93d2f4e4c3ce546d903dfc7720b58e4d7958 -> 0xcab252de675efe88ffe3b55de3e0ce14252ec7627bc746ce2211ce2ea7a81d20
0xcab252de675efe88ffe3b55de3e0ce14252ec7627bc746ce2211ce2ea7a81d20 != 0x0a72494e784c7dc1194f250237c16a0a82d8eeeb80688819b802ef49447e9741
panic: BAD transition :((
```

I haven't looked into it yet, so it's quite possible there's some dumb mistake / glaring oversight.

### CODE REVIEW

**Comment by norswap** [comment]
A possible first sanity checking step would be to restore `GenPossibleShortNodePreimage` and its two callers to see if things still work (they should, otherwise something is very wrong).

A default assumption is that there is some cache write missing somewhere. A possible process there is to go through the definition of `insert` and `delete` in `trie.go` and make sure that every node creation is appropriately registered in the cache.

The theory being that for any insert and/or removal, we're either touching nodes that we previously retrieved via `PrefetchStorage`, or new nodes that we created while doing the insertion/deletion.

Come to think of it, the above might not be quite correct, definitely worth thinking through / checking.

File: trie/node.go [hunk 1]
```diff
@@ -18,7 +18,8 @@ package trie
 
 import (
 	"fmt"
-	"io"
+	"github.com/ethereum/go-ethereum/crypto"
+"io"
```
Comment (norswap) [inline trie/node.go]: Needs fixing

File: trie/node.go [hunk 2]
```diff
@@ -47,6 +48,26 @@ type (
...
+
+func newShortNode(key []byte, val node, flags nodeFlag) *shortNode {
+    node := &shortNode{key, val, flags}
+    buf := rlp.NewEncoderBuffer(nil)
+    node.encode(buf)
+    enc := buf.ToBytes()
+    cache[crypto.Keccak256Hash(enc)] = node
+    return node
+}
+
```
Comment (norswap) [inline trie/node.go]: This will need proper documentation (same for the `cache` variable).

**Comment by norswap** [comment]
Can confirm the sanity check of restoring `GenPossibleShortNodePreimage` does indeed work.

File: core/state/statedb.go [hunk 3]
```diff
@@ -463,8 +463,6 @@ func (s *StateDB) deleteStateObject(obj *stateObject) {
 	}
 	// Delete the account from the trie
 	addr := obj.Address()
-	// Get absense proof of account in case the deletion needs the sister node.
-	oracle.PrefetchAccount(big.NewInt(s.db.BlockNumber.Int64()+1), addr, trie.GenPossibleShortNodePreimage)
```
Comment (norswap) [inline core/state/statedb.go]: This is required. The reason being that after all state updates are done, if an account is empty (all fields zero) then EIP158 mandates that it should be removed from the trie, which happens here.

A proof (= Merkle trie path) for the account has not been fetched before, only proofs against the storage root of the accounts.

However — the GenPossibleShortNodePreImage call should be unecessary (given the new node cache) and that parameter can be replaced by `nil`.

**Comment by norswap** [comment]
I've found multiple missing node creations that weren't registered (will push an updated version of the code later), but this doesn't solve the issue.

**Comment by norswap** [comment]
This is complex, so I'm going to write through my thinking.

The bug happens when we're trying to collapse a full node down to a short (aka extension) node. We do this because we deleted one of its children and it is now containing only a single child or value (but noth both).

There are two scenarios:
- the remaining child is a short node, in which case we replace the full node by a one-nibble extension of the short node
- only a value remains, or the remaining child is not a short node, in which case we replace the full node by a one-nibble-long short node with the value or child of the full node

The bug occurs before we fork into one the two scenarios though, when we try to fetch the unloaded child to check whether it is a short node or not.

The way the code works currently is to assume that either:
- the child will be present in the post-state
- the child is a short node

As far as I can tell, the assumption is completely mistaken. For instance, the child could be a full node with two children, who themselves get deleted, causing the deletion of the whole full node. Hence it would not be in the post-state.

The part about ignoring the error is also befuddling: the comment says that if we can't fetch, then the node must be a short node. However, failing the check on the next line means the node will be treated as though it was _not_ a short node.

Given that assumption however, the solution is to fetch an absence proof from the post state (this is done in `updateTrie`, in advance of the `delete` call), and to postprocess the result to include all "suffixes" of any short node included therein.

Such a "suffix" is a shorter short node whose key is a suffix of the original. The thinking there is (I think!) that a longer short node could have result from a more complex branching structure, after deletion of the "side branches".

Anyway, the "deletion" test from `make test_minigeth` fails if we remove this hacky logic. (Here's we're not even trying the reprod submitted by the hacker yet, just trying not to introduce regressions!)

What happens here is pretty simple: the `Preimage` method simply returns `nil` (but not errors/panic), causing the code to that the fork as the child was not a short node. In the case it was supposed to be, the results are obviously incorrect.

The theory as to why our changes *should* work (at least, my working copy which adds missing node caching and reintroduces the deleted `Prefetch` calls, but fetches on the pre-state, removes the short node generation) is as follows:

The node we're trying to fetch (the one whose type we're trying to determine) must either:
- exist in the pre-state
- have been created by us

If it's been created during this block transition, it should be contained in the cache. If it isn't, it should be in the pre-state.
If it is in the pre-state, it should have been fetched by one of the `PrefetchStorage` calls. In particular, the one in `updateTrie` (again, updated to fetch from the pre-state) should pull an absence proof that necessarily contains this node.

What I'm going to try next: see if can add an explicit `PrefetchStorage` for the node we're trying to fetch. That should help disambiguate whether the node exists and we're just not managing to fetch it for some reason (is the theory above wrong?) or if it's a node that we create and that we're not registering.

EDIT: Haha, of course you can't do that, there is not RPC to fetch by hash — that's why we're fetching proofs in the first place.

**Comment by norswap** [comment]
Also, it seems like the hashing computation might be busted, I'll investigate that in priority.

EDIT: I think I've fixed it, but it's rather difficult to tell without a proper testcase in place. That said, setting up such a test case is a huge effort, the simplest would be to find an existing block that accesses a value that a previous transaction wrote to. I'll try with a recent block, hoping that I don't run into the bug we're trying to fix here 😅

**Comment by norswap** [comment]
Some data on the failing deletion test case:

- hash `0x9e978fd69bda076d4bece8f7cf7b6dd684d6cfb095a76f9779240ccd98254a68`
- which is a short node with key `0x3fffb9b0249b5bfe301daa87fb1b3cd34163d5e3ca8bf77263e8b45165` and value `2`
- this is a suffix of a longer node with key `0x20efffb9b0249b5bfe301daa87fb1b3cd34163d5e3ca8bf77263e8b45165`
    - this doesn't seem like a prefix, but it's because of the hex-prefix encoding used in Ethereum: flags are packed in the first nibble, and a second nibble can be used to make the length an even number of bytes
    - In this case, the longer node as a single extra nibble (`e`) compared to the shorter node. The initial `3` and `2` in the keys are the flags, and the first `0` in the longer node is padding.

**Comment by han0110** [comment]
Hi, I think I introduced the feature (or bug 😓). I would like to clarify my thought about the hacky solution and hope it could be helpful.

The case we are dealing is a branch node is downgraded into a short node, which means we delete the second last child, and the rest cases could be:

1. The last child is not touched - The only awkwark case, we handle this by getting the absence proof in next block.
1.1. It's found to be a short node - Merge the short nodes together.
1.2. It's not found - It's either an branch node or it means the state transition is invalid.
2. The last child is updated - Since it's updated, it means we definitely have it's preimage to update it.
3. The last child is deleted - Same reason to 2.

The reason we get the absence proof in next block and call `GenPossibleShortNodePreimage` to enumerate all possible shorter short node (same suffix) is for case 1.1.

Not sure if it makes sense. Also to create some testcase, I used to try [this contract](https://gist.github.com/han0110/ec7aefc9c97bef908938b6f045054ac3) that deploys 512 children when deployed, and call them to selfdestruct when called, which will generate lots of case 1.1 and 1.2. 

> As far as I can tell, the assumption is completely mistaken. For instance, the child could be a full node with two children, who themselves get deleted, causing the deletion of the whole full node. Hence it would not be in the post-state.

If the last child is also deleted (updated to `0`), then I think we must have its preimage (case 3).

**Comment by norswap** [comment]
Hey @han0110, I appreciate the message. Don't worry about it too much, it happens!

It took me a while, but I finally understood your design yesterday. My misunderstanding centered on (1.2) indeed. Your assumption is that, because of enumerating all the possible short nodes in advance, then if the preimage is not found it must means it is a branch node.

What threw me off is that your comment is mistaken:

```
// When node is not resolved in next block's absence proof,
// it must be an extension node if the state transition is
// valid, so we ignore the error here.
```

It should say "branch node" (or "full node") instead of "extension node" :)

However, your assumption has one fatal flaw: it assumes that the resulting short node resulting from the deletion is never modified later during the block transition.

So for instance, imagine the last child has key `abc` and is at position `1` in the parent full node. The deletion causes the creation of a short node with key `1abc`. But imagine keys are inserted that break down this short node afterwards in the same block. e.g. Inserting key `1afc` in the parent of the new short node. Suddenly the `1abc` short node is no longer present in the post-state (there's a full node in the middle), and it is no longer possible to re-derive the `abc` post-node!

---

Some clarification now: all my ramblings above on this PR were not really about fixing that bug, but rather fixing a regression that occured when I tried to fix that bug.

Yesterday, I had a breakthrough on what the issue was there.

My main theory was "for all nodes to be resolved, these will either be nodes that we created during this block transition, or they will be nodes whose preimage we can get through `eth_getProof` from the pre-state upon accessing the key".

It turns out that assumption is mistaken. When trying to prefetch a key that does exist in the pre-state, the absence proof might stop at a full node that has a missing child where the key should be. In that case, only the hash of the branch's children will be provided. Hence the preimage of those children are not known from the proof.

The solution here is to detect those cases, and re-issue a fetch for any key within these children (it doesn't matter if the key exists or not,  just extend the key of the node with 0s to reach 32 bytes) , which will ensure their preimage are available.

**Comment by han0110** [comment]
> It should say "branch node" (or "full node") instead of "extension node" :)

Yes, the comment was wrong, sorry for misleading...

> it assumes that the resulting short node resulting from the deletion is never modified later during the block transition.

I assumed the trie update is only called once after all transactions have been processed, so I believe in your example it won't have issue. However, I'm not sure how this issue affects ethereum-optimism/cannon#89, so maybe I'm missing something.

---

In the project I'm working on might also need to face this issue in the future, so I really hope there is a general solution for this! So if I may I would like to ask some further following question of your solution, thanks!

> The solution here is to detect those cases, and re-issue a fetch for any key within these children (it doesn't matter if the key exists or not, just extend the key of the node with 0s to reach 32 bytes) , which will ensure their preimage are available.

Since it's using the secure trie, how do you fetch those hash of some unknown keys directly? This seems to be the root problem for this issue (if geth has such api, then we don't need to worry about anything).

**Comment by norswap** [comment]
> I assumed the trie update is only called once after all transactions have been processed, so I believe in your example it won't have issue. However, I'm not sure how this issue affects [ethereum-optimism/cannon#89](https://github.com/ethereum-optimism/cannon/issues/89), so maybe I'm missing something.

(It is indeed only called once at the end.)

If by "my exemple", you're referring to `0x9e978fd69bda076d4bece8f7cf7b6dd684d6cfb095a76f9779240ccd98254a68`, then yes, that is not the issue there.

As I explain above, the issue there is that I wasn't fetching from the post-state (only the pre-state) and by default this child node is not fetched.

However, the bad assumption I'm highlighting (assume nothing changes after deleting) has everything to do with https://github.com/ethereum-optimism/cannon/issues/89. Just check my example from above:

> So for instance, imagine the last child has key `abc` and is at position `1` in the parent full node. The deletion causes the creation of a short node with key `1abc`. But imagine keys are inserted that break down this short node afterwards in the same block. e.g. Inserting key `1afc` in the parent of the new short node. Suddenly the 1abc short node is no longer present in the post-state (there's a full node in the middle), and it is no longer possible to re-derive the abc post-node!

Applying updates once at the end doesn't save you — if you don't have the longer short node (`1abc`) in the post state, it's impossible to reconstruct the shorter short node (`abc`) from the post state.

The shorter short node must either (a) have been created in the current transition, in which case there is no problem, or (b) be present in the pre-state, where it wasn't fetched, but should be — this is the buggy case.

**Comment by han0110** [comment]
I see, will think about it. Thanks!

**Comment by norswap** [comment]
Damn, I realized I missed the second part of your message on the secure trie. That's a pretty damn good point, I need to think about that!

Btw, if I want to ask you a couple of question, is there a place I can reach out to you? I'm @norswap on telegram & twitter.

**Comment by norswap** [comment]
So, as it turns out, the issue was not at all what we thought it was.

The original code from @han0110 is fully functional, though figuring that out is no easy feat. Refer to the new `fetching-preimages.md` for an informal proof of why it works.

A few points:

- "transient nodes" was not an issue at all because of two facts:
     - Geth applies all structural trie changes in one pass, at the end of the block transition.
     - The only time we need to resolve a preimage in the trie is for `hashNode`, which represents MPT nodes that haven't been loaded yet. Nodes we created ourselves (including "transient nodes" not present in the post-state) are never represented as `hashNode`.

- The crux of @han0110 's solution is that there are preimages that we cannot fetch, but this mere impossibility is a sufficient signal to carry out the tree transformation. In particular, it means that these nodes are full nodes, which we can move by copying the `hashNode` representing them.

- The real issue was that the unicorn MIPS `Preimage` implementation didn't handle "empty" hashes correctly. These were empty files creates for the preimages we couldn't fetch, as mentionned in the previous item. Instead the implementation thought an empty byte array *was* the preimage, which failed the sanity check and paniced.

**Pull request merged**